File: C/thread.c
#include <stdio.h>
#include <pthread.h>

void* minhaThread(void* arg) {
    int id = *((int*)arg);
    printf("Olá! Eu sou a thread %d\n", id);
    return NULL;
}

int main() {
    pthread_t t1, t2;
    int id1 = 1, id2 = 2;

    // Cria duas threads
    pthread_create(&t1, NULL, minhaThread, &id1);
    pthread_create(&t2, NULL, minhaThread, &id2);

    // Espera as threads terminarem
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    printf("Todas as threads terminaram.\n");
    return 0;
}

File: Python/thread.py
import threading  # Importa o módulo 'threading' para trabalhar com múltiplas threads

contador = 0  # Variável global compartilhada entre todas as threads
lock = threading.Lock()  # Cria um objeto Lock para evitar condições de corrida

def incrementar():
    global contador  # Indica que estamos usando a variável global 'contador'
    for _ in range(100000):  # Repete 100.000 vezes
        with lock:  # Garante que apenas uma thread execute essa parte por vez
            contador += 1  # Incrementa o valor do contador de forma segura

# Cria uma lista com 5 threads, todas executando a função 'incrementar'
threads = [threading.Thread(target=incrementar) for _ in range(5)]

# Inicia todas as threads
for t in threads:
    t.start()

# Espera todas as threads terminarem antes de continuar
for t in threads:
    t.join()

# Imprime o valor final do contador (esperado: 5 * 100000 = 500000)
print("Contador final:", contador)